Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,497 @@
+  | empty =>
+    simp only [prod_empty, multinomial_empty, cast_one, sum_empty, one_mul]
+    rw [hI.dpow_zero ha]
+  | insert hi hrec =>
+    rw [prod_insert hi, hrec, ← mul_assoc, mul_comm (hI.dpow (n _) a),
+      mul_assoc, mul_dpow _ _ _ ha, ← sum_insert hi, ← mul_assoc]
+    apply congr_arg₂ _ _ rfl
+    rw [multinomial_insert hi, mul_comm, cast_mul, sum_insert hi]
+
+-- TODO : can probably be simplified using `DividedPowers.exp`
+
+/-- Lemma towards `dpow_sum` when we only have partial information on a divided power ideal
+
+See also `dpow_sum_aux'` for a more general result-/
+theorem dpow_sum_aux (dpow : ℕ → A → A)
+    (dpow_zero : ∀ {x} (_ : x ∈ I), dpow 0 x = 1)
+    (dpow_add : ∀ (n) {x y} (_ : x ∈ I) (_ : y ∈ I),
+      dpow n (x + y) = (antidiagonal n).sum fun k => dpow k.1 x * dpow k.2 y)
+    (dpow_eval_zero : ∀ {n : ℕ} (_ : n ≠ 0), dpow n 0 = 0)
+    {ι : Type*} [DecidableEq ι] {s : Finset ι} {x : ι → A} (hx : ∀ i ∈ s, x i ∈ I) (n : ℕ) :
+    dpow n (s.sum x) =
+      (Finset.sym s n).sum fun k => s.prod fun i => dpow (Multiset.count i k) (x i) := by
+  simp only [sum_antidiagonal_eq_sum_range_succ_mk] at dpow_add
+  induction s using Finset.induction generalizing n with -- a s ha ih
+  | empty =>
+    simp only [sum_empty, prod_empty, sum_const, nsmul_eq_mul, mul_one]
+    by_cases hn : n = 0
+    · rw [hn]
+      rw [dpow_zero I.zero_mem]
+      simp only [sym_zero, card_singleton, cast_one]
+    · rw [dpow_eval_zero hn, eq_comm, ← cast_zero]
+      apply congr_arg
+      rw [card_eq_zero, sym_eq_empty]
+      exact ⟨hn, rfl⟩
+  | @insert a s ha ih =>
+    have hx' : ∀ i, i ∈ s → x i ∈ I := fun i hi => hx i (mem_insert_of_mem hi)
+    simp_rw [sum_insert ha,
+      dpow_add n (hx a (mem_insert_self a s)) (I.sum_mem fun i => hx' i),
+      sum_range, ih hx', mul_sum, sum_sigma', eq_comm]
+    apply sum_bij'
+      (fun m _ => m.filterNe a)
+      (fun m _ => m.2.fill a m.1)
+      (fun m hm => mem_sigma.2 ⟨mem_univ _, _⟩)
+      (fun m hm => by
+        simp only [succ_eq_add_one, mem_sym_iff, mem_insert, Sym.mem_fill_iff]
+        simp only [mem_sigma, mem_univ, mem_sym_iff, true_and] at hm
+        intro b
+        apply Or.imp (fun h ↦ h.2) (fun h ↦ hm b h))
+      (fun m _ => m.fill_filterNe a)
+    · intro m hm
+      simp only [mem_sigma, mem_univ, mem_sym_iff, true_and] at hm
+      exact Sym.filter_ne_fill a m fun a_1 => ha (hm a a_1)
+    · intro m hm
+      simp only [mem_sym_iff, mem_insert] at hm
+      rw [prod_insert ha]
+      apply congr_arg₂ _ rfl
+      apply prod_congr rfl
+      intro i hi
+      apply congr_arg₂ _ _ rfl
+      conv_lhs => rw [← Sym.fill_filterNe a m, Sym.coe_fill]
+      simp only [Multiset.count_add, add_right_eq_self, Multiset.count_eq_zero,
+        Sym.mem_coe, Sym.mem_replicate, not_and]
+      exact fun _ => ne_of_mem_of_not_mem hi ha
+    · intro m hm
+      convert sym_filterNe_mem a hm
+      rw [erase_insert ha]
+    -- explicit arguments above rather than m.fill_filter_ne a
```

Comment:
[Nat.multinomial](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Nat/Choose/Multinomial.html#Nat.multinomial) is in mathlib!

Reply:
I don't remember what I meant. I added a note that this lemma should be golfable using the equivalence `Finset.symInsertEquiv`, but I didn't manage to do it (the issue is having to play with types, subtypes, conversion between finsets and fintypes, etc.).